### ISSUE TITLE
New Rule: no-bracket-notation-on-keyed-collections

### ISSUE THREAD

**Issue by richardtallent-erm** [eslint/eslint#16801]
### Rule details

Disallow bracket notation on keyed collection objects

### Related ECMAScript feature

Map, Set, WeakMap, and WeakSet

### What type of rule is this?

Warns about a potential problem

### Example code

```js
const m = new Map()
m["foo"] = "bar"         // Should use m.set("foo", "bar") instead
const answer = m["foo"]  // Should use m.get("foo") instead
if(m["foo"]) ...         // Should use if(m.has("foo")) instead
if(m.length > 1) ...     // Should use m.size instead

const s = new Set()
s["foo"] = true          // Should use s.add("foo") instead
const hasFoo = s["foo"]  // Should use s.has("foo") instead
if(s.length > 1) ...     // Should use s.size instead
```

### Why should this rule be in the core instead of a plugin?

`Map`, `Set`, `WeakMap`, and `WeakSet` are standard, built-in JavaScript objects. They don't support using bracket notation semantics to access members of their collection, and instead have their own methods for setting, getting, and testing the presence of elements.

This can be confusing to developers who are more familiar with using plain objects as dictionaries. Also, when converting code from plain objects or arrays to keyed collections, or maintaining the code using them later, bracket notation can slip by in the code. These can result in confusing "undefined" results.

It's *possible* someone would want to add custom properties of one of these objects. However, doing so using bracket notation should be strongly discouraged, since if it's not an error, it smells like one.

Note that there is one exception -- calling the iterator function:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/@@iterator
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/@@iterator

Since these are only indexing a specific static Symbol, hopefully those specific calls can be excluded from the rule.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

**Comment by mdjermanovic** [comment]
Hi @richardtallent-erm, thanks for the issue!

Rules that relate to object types would require type information that we don't have in the core. Also, this looks to me more like a check that belongs to type checking rather than linting. You could run `tsc` on JS files to catch these errors, here's a [TS Playground demo](https://www.typescriptlang.org/play?filetype=js#code/MYewdgzgLgBAtjAvDMBTA7jAsgQwA4AUAlAFBwDaARAGYgiUC6SMlARjgE6Uw++8D0-GAGUAFiACuAGwAmMCRFTwAdIqgEadSgBoW7LkRgBLSFFQ4ZJUKZg5I6VB2YVN9JjEEjx0uQqVxlAHNUdVdKQxNoc0sjagIXWjdDAG8AXz4MzzFJWXlFYziA0RwIDUTwiNNoklj45SlUMECoURgAPhgARhT0vizvXL8VCCMALyVIswsSK3BoGAhmNExhEOISCCpypmQoDgklDL6hbJ88pQhlCxkyrUqo6et54ogAMTpmTbD3fpzffMuL1u9HuUxicUuDSaLXaXR6mROA3+F1UYwmVWmQA) with the examples and they're all marked as errors.
